File: src/test/java/cube8540/oauth/authentication/oauth/error/OAuth2ExceptionTranslatorTest.java
package cube8540.oauth.authentication.oauth.error;

import cube8540.oauth.authentication.oauth.token.domain.OAuth2AccessTokenNotFoundException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.http.CacheControl;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.oauth2.core.OAuth2Error;
import org.springframework.security.oauth2.core.OAuth2ErrorCodes;
import org.springframework.web.HttpRequestMethodNotSupportedException;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@DisplayName("OAuth2 기본 에러 변환기 테스트")
class OAuth2ExceptionTranslatorTest {

    private OAuth2ExceptionTranslator translator;

    @BeforeEach
    void setup() {
        this.translator = new OAuth2ExceptionTranslator();
    }

    @Test
    @DisplayName("OAuth2 클라이언트 인증 관련 예외 변환")
    void translateOAuth2ClientAuthenticationException() {
        OAuth2Error error = new OAuth2Error(OAuth2ErrorCodes.INVALID_CLIENT);
        AbstractOAuth2AuthenticationException exception = mock(AbstractOAuth2AuthenticationException.class);

        when(exception.getStatusCode()).thenReturn(401);
        when(exception.getError()).thenReturn(error);

        ResponseEntity<OAuth2Error> result = translator.translate(exception);
        assertEquals(HttpStatus.UNAUTHORIZED, result.getStatusCode());
        assertEquals(error, result.getBody());
        assertEquals(CacheControl.noStore().getHeaderValue(), result.getHeaders().getCacheControl());
        assertEquals("no-cache", result.getHeaders().getPragma());
    }

    @Test
    @DisplayName("OAuth2 지원 되지 인증 방식 관련 예외 변환")
    void translateOAuth2NotSupportedAuthenticationMethodException() {
        OAuth2Error error = new OAuth2Error(OAuth2ErrorCodes.UNSUPPORTED_GRANT_TYPE);
        AbstractOAuth2AuthenticationException exception = mock(AbstractOAuth2AuthenticationException.class);

        when(exception.getStatusCode()).thenReturn(400);
        when(exception.getError()).thenReturn(error);

        ResponseEntity<OAuth2Error> result = translator.translate(exception);
        assertEquals(HttpStatus.BAD_REQUEST, result.getStatusCode());
        assertEquals(error, result.getBody());
        assertEquals(CacheControl.noStore().getHeaderValue(), result.getHeaders().getCacheControl());
        assertEquals("no-cache", result.getHeaders().getPragma());
    }

    @Test
    @DisplayName("지원 되지 않는 메소드 타입에 관한 예외 변환")
    void translateNotSupportedMethodException() {
        HttpRequestMethodNotSupportedException exception = mock(HttpRequestMethodNotSupportedException.class);

        ResponseEntity<OAuth2Error> result = translator.translate(exception);
        assertEquals(HttpStatus.METHOD_NOT_ALLOWED, result.getStatusCode());
        assertEquals(CacheControl.noStore().getHeaderValue(), result.getHeaders().getCacheControl());
        assertEquals("no-cache", result.getHeaders().getPragma());
    }

    @Test
    @DisplayName("클라이언트 등록 관련 예외 변환")
    void translateClientRegistrationException() {
        OAuth2ClientRegistrationException exception = mock(OAuth2ClientRegistrationException.class);

        ResponseEntity<OAuth2Error> result = translator.translate(exception);
        assertEquals(HttpStatus.UNAUTHORIZED, result.getStatusCode());
        assertEquals(CacheControl.noStore().getHeaderValue(), result.getHeaders().getCacheControl());
        assertEquals("no-cache", result.getHeaders().getPragma());
    }

    @Test
    @DisplayName("엑세스 토큰을 찾을 수 없는 예외 변환")
    void translateOAuth2AccessTokenNotFoundException() {
        OAuth2AccessTokenNotFoundException exception = mock(OAuth2AccessTokenNotFoundException.class);

        ResponseEntity<OAuth2Error> result = translator.translate(exception);
        assertEquals(HttpStatus.BAD_REQUEST, result.getStatusCode());
        assertEquals(CacheControl.noStore().getHeaderValue(), result.getHeaders().getCacheControl());
        assertEquals("no-cache", result.getHeaders().getPragma());
    }

    @Test
    @DisplayName("정의 되지 않은 예외 변환")
    void translateNotDefinedException() {
        Exception exception = mock(Exception.class);

        ResponseEntity<OAuth2Error> result = translator.translate(exception);
        assertEquals(HttpStatus.INTERNAL_SERVER_ERROR, result.getStatusCode());
        assertEquals(CacheControl.noStore().getHeaderValue(), result.getHeaders().getCacheControl());
        assertEquals("no-cache", result.getHeaders().getPragma());
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/error/OAuth2ExceptionTranslatorTest.kt
package cube8540.oauth.authentication.oauth.error

import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.security.oauth2.core.OAuth2ErrorCodes
import org.springframework.web.HttpRequestMethodNotSupportedException
import java.util.stream.Stream

class OAuth2ExceptionTranslatorTest {

    private val translator = OAuth2ExceptionTranslator()

    @ParameterizedTest
    @MethodSource(value = ["abstractOAuth2AuthenticationException"])
    fun `translate abstract oauth2 authentication exception`(exception: AbstractOAuth2AuthenticationException) {
        val result = translator.translate(exception)

        assertThat(result.statusCode).isEqualTo(HttpStatus.valueOf(exception.statusCode))
        assertThat(result.body).isEqualTo(exception.error)
        assertThat(result.headers[HttpHeaders.CACHE_CONTROL]).isEqualTo(listOf("no-store"))
        assertThat(result.headers.pragma).isEqualTo("no-cache")
    }

    @ParameterizedTest
    @MethodSource(value = ["httpRequestMethodNotSupportedException"])
    fun `translate http request method not supported exception`(exception: HttpRequestMethodNotSupportedException) {
        val result = translator.translate(exception)

        assertThat(result.statusCode).isEqualTo(HttpStatus.METHOD_NOT_ALLOWED)
        assertThat(result.body!!.errorCode).isEqualTo("method_not_allowed")
        assertThat(result.body!!.description).isEqualTo(exception.message)
        assertThat(result.headers[HttpHeaders.CACHE_CONTROL]).isEqualTo(listOf("no-store"))
        assertThat(result.headers.pragma).isEqualTo("no-cache")
    }

    @ParameterizedTest
    @MethodSource(value = ["oauth2ClientRegistrationException"])
    fun `translate oauth2 client registration exception`(exception: OAuth2ClientRegistrationException) {
        val result = translator.translate(exception)

        assertThat(result.statusCode).isEqualTo(HttpStatus.UNAUTHORIZED)
        assertThat(result.body!!.errorCode).isEqualTo(OAuth2ErrorCodes.UNAUTHORIZED_CLIENT)
        assertThat(result.body!!.description).isEqualTo(exception.message)
        assertThat(result.headers[HttpHeaders.CACHE_CONTROL]).isEqualTo(listOf("no-store"))
        assertThat(result.headers.pragma).isEqualTo("no-cache")
    }

    @ParameterizedTest
    @MethodSource(value = ["oauth2AccessTokenRegistrationException"])
    fun `translate oauth2 access token registration exception`(exception: OAuth2AccessTokenRegistrationException) {
        val result = translator.translate(exception)

        assertThat(result.statusCode).isEqualTo(HttpStatus.BAD_REQUEST)
        assertThat(result.body!!.errorCode).isEqualTo(OAuth2ErrorCodes.INVALID_REQUEST)
        assertThat(result.body!!.description).isEqualTo(exception.message)
        assertThat(result.headers[HttpHeaders.CACHE_CONTROL]).isEqualTo(listOf("no-store"))
        assertThat(result.headers.pragma).isEqualTo("no-cache")
    }

    @Test
    fun `translate another exception`() {
        val exception = Exception("exception")
        val result = translator.translate(exception)

        assertThat(result.statusCode).isEqualTo(HttpStatus.INTERNAL_SERVER_ERROR)
        assertThat(result.body!!.errorCode).isEqualTo(OAuth2ErrorCodes.SERVER_ERROR)
        assertThat(result.body!!.description).isEqualTo(exception.message)
        assertThat(result.headers[HttpHeaders.CACHE_CONTROL]).isEqualTo(listOf("no-store"))
        assertThat(result.headers.pragma).isEqualTo("no-cache")
    }

    private fun abstractOAuth2AuthenticationException() = Stream.of(
        Arguments.of(
            mockk<AbstractOAuth2AuthenticationException> {
                every { statusCode } returns 400
                every { error } returns mockk()
            }
        ),
        Arguments.of(
            mockk<AbstractOAuth2AuthenticationException> {
                every { statusCode } returns 404
                every { error } returns mockk()
            }
        ),
        Arguments.of(
            mockk<AbstractOAuth2AuthenticationException> {
                every { statusCode } returns 500
                every { error } returns mockk()
            }
        )
    )

    private fun httpRequestMethodNotSupportedException() = Stream.of(
        Arguments.of(HttpRequestMethodNotSupportedException("method not supported"))
    )

    private fun oauth2ClientRegistrationException() = Stream.of(
        Arguments.of(OAuth2ClientRegistrationException("client registration exception"))
    )

    private fun oauth2AccessTokenRegistrationException() = Stream.of(
        Arguments.of(OAuth2AccessTokenRegistrationException("token registration exception"))
    )
}